server/matchmaker/algorithm.py

```diff
@@ -151,22 +120,45 @@ def find(self) -> Dict[Search, Search]:
 
 @with_logger
 class Matchmaker(object):
-    def __init__(self, searches: List[Search]):
+    def __init__(self, searches: Iterable[Search]):
         self.searches = searches
         self.matches: Dict[Search, Search] = {}
 
     def find(self) -> List[Match]:
         self._logger.debug("Matching with stable marriage...")
-        self.matches.update(StableMarriage(self.searches).find())
+        searches = [s for s in self.searches if s is not None]
+        if len(searches) < 30:
+            ranks = _MatchingGraph.build_full(searches)
+        else:
+            ranks = _MatchingGraph.build_fast(searches)
+        self.matches.update(StableMarriage().find(ranks))
+
+        self._register_unmatched_searches()
 
         remaining_searches = [
             search for search in self.searches if search not in self.matches
         ]
         self._logger.debug("Matching randomly for remaining newbies...")
-        self.matches.update(RandomlyMatchNewbies(remaining_searches).find())
+        self.matches.update(RandomlyMatchNewbies().find(remaining_searches))
 
         return self._remove_duplicates()
 
+    def _register_unmatched_searches(self):
+        """
+        Tells all unmatched searches that they went through a failed matching
+        attempt.
+        """
+        unmatched_searches = filter(
+            lambda search: search not in self.matches,
+            self.searches
+        )
```

Comment:
Imho, we should use one of this: list comprehension or this map functions.
But this is more on discussion. In fact there is no reason to not use both.